ADR/0039-spdx-support.md

```diff
@@ -0,0 +1,146 @@
+| ]                               | ]                                                          |
+
+
+#### Formulations
+CycloneDX formulations describe how the container was manufactured. In SPDX, Relationship elements can be used for the same purpose. All elements in SPDX have SPDXID attribute which is an element identifier unique in the whole SBOM document. Relationship element describes relation between two elements using their SPDXID and relationship type. Relationship type BUILD_TOOL_OF can be used to express the relationship of packages which were used to build the container.
+
+| SPDX Attribute                  | CycloneDX Attribute                                        |
+|---------------------------------|------------------------------------------------------------|
+| Formulations.components = [{}]  | Relationships = [{                                         |
+|                                 |  spdxElementId = <A-PACKAGE-ID>                            |
+|                                 |  relationshipType=BUILD_TOOL_OF                            |
+|                                 |  relatedSpdxElement=<ROOT-DOCUMENT-ID>                     |
+|                                 | }]                                                         |
+
+
+#### Metadata.tools
+The CycloneDX metadata.tools sub attributes that we are mostly interested in are the vendor and name elements. Information about the creation of the SPDX document can be stored into creationInfo. CreationInfo.creators element is basically a list of strings. There’s a vague specification about how it should be structured in the standard. Strings should be formatted in the following way: “<Attribute>: <Value>”. For example vendor should be stored as “Vendor: <vendor>”
+
+| SPDX Attribute                                 | CycloneDX Attribute                               |
+|------------------------------------------------|---------------------------------------------------|
+| Metadata.tools = [{“vendor”: “X”, “name”: “Y”] | CreationInfo.creators = [“Vendor: X”, “Tool: Y”]  |
+
+
+#### Merging SPDX
+##### Packages
+Packages of two SPDX documents can be merged together as a concatenation of two lists. In cycloneDX component elements can have only a single purl attribute, therefore component elements representing packages with the same name and version but with different purl have to be stored as multiple elements. CycloneDX package elements can bear multiple purls. Therefore multiple cycloneDX components can be squashed together into single SPDX package element with purls concatenated into a single list. Following rules are applied to packages merging process:
+- Packages with the same package name and versionInfo are squashed into single package element
+- Package with the same package name and versionInfo set to None (or empty) is squashed with package with the same name and non-empty versionInfo
```

Comment:
I think merging components as described here is not correct. Having the same name and version is not sufficient indication that two components are the same component.

Consider two "same" components except one is an npm package and the other a python package
Or two identical-looking python packages, except one is from PyPI and the other is a random project with the same name and version hosted only on github.

IMO it's almost never correct to merge two components if their purls differ

Reply:
Can you put here real world example? I'm only thinking of lets say "RPM" which can be both rpm package and python package. But eventually they are referring to the same component. For package which is installed github and pip, if it's same package it still does make sense to put it in one component. Still, we're merging sboms of one container produced by different tools or sboms of base image and the container.
Real world example of that situation is there's lets say python package X reported with no version by one tool and with some version reported by another tool. Note that even the name doesn't necessarily have to be same as one tool can report PyYAML and another pyyaml for example. I think the goal here is to produce sboms which doesn't have 4 different pyaml SPDX packages reported.
Note that it would help if cachi2 could generate components/packages with identifier marking the package as python, rpm, npm and so on. That would help in the merging process.
Merging just based on purl is nonsense as you in will work with released images having SBOM packages with multiple purls with with release-specific attributes like "&repo=xxx". So no purl in build phase will match with those.

Reply:
I think it might be useful to show an example here. In SPDX, you might easily have something like this:
```
"packages": [
  {
    packageA,
    "externalRefs": [
      {
        "referenceCategory": "PACKAGE-MANAGER",
        "referenceType": "purl",
        "referenceLocator": purlA
      },
      {
        "referenceCategory": "PACKAGE-MANAGER",
        "referenceType": "purl",
        "referenceLocator": purlB
      }
    ]
  },
  ...
],
```

That is, a single package has multiple purls. Think of a container image which gets shipped to multiple repositories, such as ubi9. Or, really, any container image which bears multiple tags. Here's a hand-crafted example: https://github.com/RedHatProductSecurity/security-data-guidelines/blob/main/sbom/examples/container_image/release/ubi9-micro-container-9.4-6.1716471860.spdx.json

What this section is saying is that, to express this in CycloneDX 1.4, the only feasible option is to encode it as multiple components, identical except for their purl. (FWIW CycloneDX 1.6 provides the additional option of providing multiple forms of identity evidence: https://cyclonedx.org/docs/1.6/json/#components_items_evidence_identity_oneOf_i0_items_confidence)

So the equivalent CycloneDX 1.4 would need to look like this:
```
"components": [
  {
    "bom-ref": packageA.bom_ref
    packageA...
    "purl": purlA
  },
  {
    "bom-ref": packageA.bom_ref + "-2",
    packageA...
    "purl": purlB
  },
  ...
]
```

So, the trick is to recognise when these multiple CycloneDX components should be merged into a single SPDX package. One way to do this which might be less controversial is only to merge when the checksums are identical?

But I wonder -- would it cause a problem if they were encoded as multiple SPDX packages?

Reply:
Cachi2 doesn't populate checksum field for sboms. For further processing I think multiple spdx package could  present additional difficulty. Like if you want to check if container contains specific package with a vulnerability you would have to iterate over several package. Also without any reduction of packages having containers like A->B->C (A is base image of B and so on) could lead to at least 6 spdx packages representing one real package (1 package cachi2 + 1 syft x number of purl variations with repo,arch,etc.)

Reply:
> Note that it would help if cachi2 could generate components/packages with identifier marking the package as python, rpm, npm and so on.

The type is part of the purl: `pkg:{type}/{name}@{version}`

Cachi2 merges by the full purl, because in cachi2 different purls = different components (or at least components where you can't be certain they're the same). But for merging in other contexts, like release SBOMs, using only a subset of the purl attributes would be reasonable.

Is your goal here to come up with a generic merging solution? It may be more feasible to tailor it to specific use cases. Merging SBOMs in general and doing it well is, in my opinion, almost impossible

Reply:
So can we agree on merging based on type name and version of purl?

Reply:
For merging cachi2 SBOMs with other cachi2 SBOMs - no, it should continue merging by the full purl.

For merging cachi2 SBOMs with Syft SBOMs, yes that sounds reasonable. Although I believe the current merge script has some more specialized syft-specific logic already.

For generic merging, yes that sounds reasonable